spec/system/tech_shop_spec.rb

```diff
@@ -37,150 +39,142 @@
+      expect(products_before_clear).not_to be_empty
+
+      Product.clear_index!(true)
+
+      products_after_clear = Product.raw_search('')['hits']
+      expect(products_after_clear).to be_empty
+      Product.reindex!(MeiliSearch::Rails::IndexSettings::DEFAULT_BATCH_SIZE, true)
+
+      products_after_reindex = Product.raw_search('')['hits']
+      expect(products_after_reindex).not_to be_empty
+      expect(products_before_clear).to eq(products_after_reindex)
+    end
   end
 
   describe 'basic searching' do
     it 'finds the iphone' do
       results = Product.search('iphone')
-      expect(results.size).to eq(1)
-      expect(results).to include(@iphone)
+      expect(results).to contain_exactly(@iphone)
     end
 
     it 'searches case insensitively' do
       results = Product.search('IPHONE')
-      expect(results.size).to eq(1)
-      expect(results).to include(@iphone)
+      expect(results).to contain_exactly(@iphone)
     end
 
     it 'finds all amazon products' do
       results = Product.search('amazon')
-      expect(results.size).to eq(3)
-      expect(results).to include(@android, @samsung, @motorola)
+      expect(results).to contain_exactly(@android, @samsung, @motorola)
     end
 
     it 'finds all "palm" phones with wildcard word search' do
       results = Product.search('pal')
-      expect(results.size).to eq(2)
-      expect(results).to include(@palmpre, @palm_pixi_plus)
+      expect(results).to contain_exactly(@palmpre, @palm_pixi_plus)
     end
 
     it 'searches multiple words from the same field' do
       results = Product.search('palm pixi plus')
-      expect(results.size).to eq(1)
-      expect(results).to include(@palm_pixi_plus)
+      expect(results).to contain_exactly(@palm_pixi_plus)
     end
 
     it 'finds using phrase search' do
       results = Product.search('coco "palm"')
-      expect(results.size).to eq(1)
-      expect(results).to include(@palm_pixi_plus)
+      expect(results).to contain_exactly(@palm_pixi_plus)
     end
 
     it 'narrows the results by searching across multiple fields' do
       results = Product.search('apple iphone')
-      expect(results.size).to eq(2)
-      expect(results).to include(@iphone)
+      expect(results).to include(@iphone, @macbook)
     end
 
     it 'does not search on non-indexed fields' do
-      results = Product.search('features')
-      expect(results.size).to eq(0)
+      expect(Product.search('features')).to be_empty
     end
 
-    it 'deletes the associated record' do
+    it 'deletes associated document on #destroy' do
       ipad = Product.create!(name: 'ipad', href: 'apple', tags: ['awesome', 'great battery'],
                              description: 'Big screen')
 
       ipad.index!(true)
       results = Product.search('ipad')
-      expect(results.size).to eq(1)
+      expect(results).to contain_exactly(ipad)
 
       ipad.destroy
-      results = Product.search('ipad')
-      expect(results.size).to eq(0)
```

Comment:
Fun fact: this actually tested nothing.

As one of the later tests asserts, if there is a document in the ms index that cannot be found in AR, it won't be returned by `search` (although it is returned by `raw_search`).

This is why this never caused any errors despite the obvious-looking race condition between this assertion and the above `ipad.destroy` line.